lib/good_job/notifier.rb

```diff
@@ -0,0 +1,97 @@
+require 'concurrent/atomic/atomic_boolean'
+
+module GoodJob
+  class Notifier
+    CHANNEL = 'good_job'.freeze
+    WAIT_INTERVAL = 1
+
+    POOL_OPTIONS = {
+      min_threads: 0,
+      max_threads: 1,
+      auto_terminate: true,
+      idletime: 60,
+      max_queue: 1,
+      fallback_policy: :discard,
+    }.freeze
+
+    cattr_reader :instances, default: [], instance_reader: false
+
+    def self.notify(message)
+      connection = ActiveRecord::Base.connection
+      connection.exec_query <<~SQL
+        NOTIFY #{CHANNEL}, #{connection.quote(message.to_json)}
+      SQL
+    end
+
+    attr_reader :recipients
+
+    def initialize(*recipients)
+      @recipients = Concurrent::Array.new(recipients)
+      @pool = Concurrent::ThreadPoolExecutor.new(POOL_OPTIONS)
+      @listening = Concurrent::AtomicBoolean.new(false)
+
+      self.class.instances << self
+
+      listen
+    end
+
+    def listening?
+      @listening.true?
+    end
+
+    def shutdown(wait: true)
+      return unless @pool.running?
+
+      @pool.shutdown
+      @pool.wait_for_termination if wait
+    end
+
+    def shutdown?
+      !@pool.running?
+    end
+
+    def listen
+      future = Concurrent::Future.new(args: [@recipients, @pool, @listening], executor: @pool) do |recipients, pool, listening|
+        Rails.application.reloader.wrap do
+          connection = ActiveRecord::Base.connection
+          conn = connection.instance_variable_get(:@connection)
```

Comment:
This can use `raw_connection` instead https://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/AbstractAdapter.html#method-i-raw_connection